Refactor registration of models in modelsmodule

clinssen opened the pull request:
This PR refactors the code that registers models into the kernel in `modelsmodule.h/.cpp`, into something that is functionally equivalent but eliminates about half of the lines of code that pertain to adding synapses. This is because each synapse type was explicitly added twice: each with its own index data type (`TargetIdentifierPtrRport` and `TargetIdentifierIndex`). I moved out all the `register_connection_model()` calls out to a templated function, and then call the function twice, once with each data type given above.

Also removed a large amount of source code comments that did not belong here.

heplesser commented on models/modelsmodule.h:
```diff
@@ -49,6 +51,10 @@ class ModelsModule : public SLIModule
+
+protected:
+  template < typename ConnectionT >
+  void register_connection_models( std::string name_postfix = "" );
```
Please add Doxygen documentation for the method. 
And is this method specific to modelsmodule? Wouldn't it be useful also for other modules defining synapses and should thus rather be provided by `ModelManager`?

heplesser replied:
- As it is a private method, add `_` at the end of the method name.
- Concerning the docstring:
    - No need to write that it is for internal use, the fact it is `private` makes that clear.
    - Docstring should start with a single line short description
    - Add a Note to make clear that for any new connection model, registration must be added in this method.

heplesser commented on models/modelsmodule.cpp:
```diff
@@ -297,72 +297,6 @@ ModelsModule::init( SLIInterpreter* )
     "volume_transmitter" );
 
   // Create voltmeter as a multimeter pre-configured to record V_m.
-  /*BeginDocumentation
```
Where does the `voltmeter` documentation go?

heplesser commented on models/modelsmodule.cpp:
```diff
@@ -432,48 +366,12 @@ ModelsModule::init( SLIInterpreter* )
-  // register synapses
-
-  /* BeginDocumentation
-     Name: static_synapse_hpc - Variant of static_synapse with low memory
```
Where does this documentation go?

jougs replied:
@heplesser: the documentation is now [here](https://github.com/nest/nest-simulator/pull/1069/files#diff-4d73c5bddaa99f5c5da31c447c96dc33R185).

heplesser commented on models/modelsmodule.cpp:
```diff
@@ -432,48 +366,12 @@ ModelsModule::init( SLIInterpreter* )
+  // register all connection models: once for the normal indexing type and
+  // once for the more efficient "HPC" indexing type
+  register_connection_models< TargetIdentifierPtrRport >();
+  register_connection_models< TargetIdentifierIndex >( "_hpc" );
```
This makes me almost wonder if we shouldn't place the suffix information in the `TargetIdentifier...` classes, to make sure that they are consistent throughout. With the current setup, users could create any kind of suffix-TI combinations.

heplesser replied:
If this is challenging to implement, I suggest to create a new issue to remind us to consider further refactoring later and for now use the code you propose here.

clinssen commented:
The NEST Open Developer's Meeting of November 12th, 2018, decided that the "voltmeter" is a useful concept (a kind of "syntactic sugar") and will remain supported. Where the initialization and documentation (in the form of a source code comment block) should go, however, is not yet clear. Will await further comments from @jougs on this issue while I address the other comments made by @heplesser.

gtrensch commented on models/modelsmodule.h:
```diff
@@ -49,6 +51,10 @@ class ModelsModule : public SLIModule
   const std::string name( void ) const;
   const std::string commandstring( void ) const;
+
+protected:
```
Shouldn't it be private? There is no inheritance, all happens in the ModelsModule class.

jougs commented:
@clinssen: what is the status of this? Thanks!

heplesser commented on models/voltmeter.cpp:
```diff
@@ -0,0 +1,60 @@
+/*
+ *  multimeter.cpp
```
`voltmeter.cpp`

heplesser commented on models/voltmeter.cpp:
```diff
@@ -0,0 +1,60 @@
+Voltmeter::Voltmeter()
+  : Multimeter()
+{
+    //record_from_.push_back( names::V_m );
```
Remove commented code.

heplesser commented on models/voltmeter.cpp:
```diff
@@ -0,0 +1,60 @@
+{
+  /*DictionaryDatum vmdict;
+  n.get_status(vmdict);
+  assert(vmdict[names::record_from] == names::V_m.toString());*/
```
Remove commented code.

heplesser commented on models/voltmeter.cpp:
```diff
@@ -0,0 +1,60 @@
+  assert (p.record_from_.size() == 1 
+          && p.record_from_[0] == names::V_m );
+  interval_.calibrate();
+}*/
```
Remove commented code.

heplesser commented on models/voltmeter.h:
```diff
@@ -0,0 +1,123 @@
+/*
+ *  multimeter.h
```
`voltmeter.h`

heplesser commented on models/modelsmodule.h:
```diff
@@ -27,6 +27,8 @@
 
 // Includes from sli:
+#include "connector_model.h"
+#include "genericmodel.h"
```
From the changes in this file, I don't see why you would need to add these two includes here. Remove if not needed, or move to `modelsmodule.cpp` if needed there. They are also misplaced here under heading `Includes from sli`.

jougs commented:
@clinssen and I will work on this when I'm back in Jülich in approx. two weeks.

jougs commented:
As discussed offline, the `enum` type safety problems can be solved by following [this approach](http://blog.bitwigglers.org/using-enum-classes-as-type-safe-bitmasks/). @clinssen will add this as a general macro-based solution in `libnestutil`. Probably something like this can also be used by @heplesser to express the simulator states in #556. @gtrensch and @terhorstd are also interested in this.

ikitayama commented:
Hi,
If you guys can share the NEST3 branch that has the redactor codebase,
would you point me to it?

On Tue, Apr 9, 2019 at 19:05 Jochen Martin Eppler <notifications@github.com>
wrote:

> As discussed offline, the enum type safety problems can be solved by
> following this approach
> <http://blog.bitwigglers.org/using-enum-classes-as-type-safe-bitmasks/>.
> @clinssen <https://github.com/clinssen> will add this as a general
> macro-based solution in libnestutil. Probably something like this can
> also be used by @heplesser <https://github.com/heplesser> to express the
> simulator states in #556
> <https://github.com/nest/nest-simulator/issues/556>. @gtrensch
> <https://github.com/gtrensch> and @terhorstd
> <https://github.com/terhorstd> are also interested in this.
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/nest/nest-simulator/pull/1069#issuecomment-481188284>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AA6w5yvPNrOfcTNx-vutYInLu7HiR2Bkks5vfGXvgaJpZM4YVIOb>
> .
>

clinssen commented:
Dear @ikitayama, you may find yourself in the wrong thread. Could you open a separate issue (or email to nest_user) for this issue? Thanks!

clinssen commented:
Another potential implementation of bitfields (due to @terhorstd) for consideration as an alternative: https://gist.github.com/terhorstd/e882e6cef744f10b44f87473343b4b4f

ikitayama commented:
> Dear @ikitayama, you may find yourself in the wrong thread. Could you open a separate issue (or email to nest_user) for this issue? Thanks!

Well, I was actually interested in your refactoring work discussed in this PR #1069. If this isn't
merged to the mainline, in the meantime, I'll do testing on your branch. Thanks!

gtrensch commented on nestkernel/model_manager.h:
```diff
@@ -179,26 +180,26 @@ class ModelManager : public ManagerInterface
+    const bool supports_wfr=false,
+    const bool requires_symmetric=false,
+    const bool requires_clopath_archiving=false);
 
```
The point with Boolean arguments is that they are an indicator that a function does more than one thing, which should be avoided for a clean coding style.

clinssen commented:
@gtrensch: with gcc8 and other fixes having been merged into Travis CI (#1267, #1310), this PR is now ready for re-review. If you have a spare moment, would you have a look? I will be happy to walk through it together. The main change since last time is that a C++ unit test has been added for the bitfield enum class.

jougs commented:
@gtrensch, can you please have another look at this one? Thanks!

jougs commented on models/voltmeter.cpp:
```diff
@@ -0,0 +1,44 @@
+namespace nest
+{
+Voltmeter::Voltmeter()
+  : Multimeter()
```
Since the merge of #1345, both `Multimeter` and `Voltmeter` start with a lower case letter. Please adapt all occurrences accordingly.

jougs commented on nestkernel/model_manager.h:
```diff
@@ -178,32 +179,24 @@ class ModelManager : public ManagerInterface
   void register_connection_model( const std::string& name,
-    const bool requires_symmetric = false,
-    const bool requires_clopath_archiving = false );
+    const Register_Connection_Model_Flags flags = default_connection_model_flags );
```
Could `Register_Connection_Model_Flags` please be just CamelCase without the underbars?

jougs commented on models/modelsmodule.cpp:
```diff
@@ -380,214 +293,39 @@ ModelsModule::init( SLIInterpreter* )
+  register_secondary_connection_model< RateConnectionDelayed >(
+    "rate_connection_delayed", Register_Connection_Model_Flags::HAS_DELAY );
+  register_secondary_connection_model< DiffusionConnection >(
+    "diffusion_connection", Register_Connection_Model_Flags::SUPPORTS_WFR );
```
:heart: :heart: :heart:

clinssen commented:
Thanks for the praise! The new commits should address all of your comments.

jougs commented:
@heplesser: could you please indicate if the recent changes address your concerns? Thanks!

heplesser commented on models/CMakeLists.txt:
```diff
@@ -77,6 +77,7 @@ set( models_sources
     mip_generator.h mip_generator.cpp
     modelsmodule.h modelsmodule.cpp
     multimeter.h multimeter.cpp
+    voltmeter.h voltmeter.cpp
```
Please put this in proper alphabetical order.

heplesser commented on libnestutil/enum_bitfield.h:
```diff
@@ -0,0 +1,127 @@
+ *
+ * .. code-block:: C++
+ *
+ *    enum class My_Flags : unsigned
```
I am not too fond of mixing camelcasing and underscores. How about `MyFlags`?

jougs replied:
Oh! I must have missed this. This definitely needs to go away!

heplesser commented on libnestutil/enum_bitfield.h:
```diff
@@ -0,0 +1,127 @@
+ *
+ *    My_Flags my_flags = My_Flags::FIRST_FLAG | My_Flags::FOURTH_FLAG;
+ *    my_flags |= My_Flags::THIRD_FLAG;
+ *    if (enumFlagSet(my_flags, My_Flags::FOURTH_FLAG))
```
Spaces inside the `()`.

heplesser commented on libnestutil/enum_bitfield.h:
```diff
@@ -0,0 +1,127 @@
+ * .. code-block:: C++
+ *
+ *    template <>
+ *    struct EnableBitMaskOperators< My_Flags >
```
Could you add some more information on what this does, why it is needed? Why can't I go straight from the first code block above to the last one below?

Is there any risk that use of `static` gobal variables could cause problem in threaded situations? I assume not, but would like it confirmed.

clinssen replied:
The magic happens entirely in the compiler! I'm not sure whether the EnableBitMaskOperators struct is ever even instantiated at runtime.

heplesser commented on libnestutil/enum_bitfield.h:
```diff
@@ -0,0 +1,127 @@
+ *
+ *    enum class My_Flags : unsigned
+ *    {
+ *        FIRST_FLAG = 1 << 0,
```
I could use any names for the flags here, correct?

clinssen replied:
Correct, we could alternatively do `FOO_FLAG`, `XYZZY_FLAG`, ...

heplesser commented on models/modelsmodule.h:
```diff
@@ -29,10 +29,17 @@
 // Includes from sli:
 #include "slimodule.h"
 
+#include "nest.h"
```
Add  `// Includes from nestkernel:` on line above and move section before includes from SLI

heplesser commented on nestkernel/model_manager_impl.h:
```diff
@@ -79,54 +81,59 @@ ModelManager::register_preconf_node_model( const Name& name,
 
-template < typename ConnectionT, template < typename > class ConnectorModelT >
+// we're templating this function with a class template, e.g.
+// BernoulliConnection< targetidentifierT >
```
Could you move this information to the Doxygen comment for the method in the h-file?

heplesser commented on nestkernel/nest.h:
```diff
@@ -55,6 +56,45 @@ void reset_network();
 
 void register_logger_client( const deliver_logging_event_ptr client_callback );
+
+enum class RegisterConnectionModelFlags : unsigned
```
I would prefer `unsigned int` to just `unsigned`

clinssen commented:
Thanks for the feedback! The new commits should resolve all your remaining comments.

clinssen commented:
Whoops! I amended the last commit and did a force push, to keep the binary blob out of the repository's history.
